BLOCKED Issues/808 seed demo data on server startup

YaroslavLitvinov opened the pull request:
### When it's done it will close #808, #844

Note: _Work on PR stopped, since it touches architectural aspects that need to be discussed before resuming work._

* api-seed crate: Seeds fake entities into embucket if enabled with cmd line arg `--seed-data-variant`.  Planned to create a separate executable for doing the same.
* api-structs crate: Structs traits decoupling and refactoring, split on derive macros level. 
* api-client-rest crate (Rust SDK): 
  - High level operations on DB, create volumes, create schemas, (no tables and table data yet)
  - *requests* generic nested module supports embucket authentication, run queries with session support, generic rest calls
* Tests OK  

### This PR is also used a place for experiments related to #844

rampage644 commented:
@YaroslavLitvinov Let's have a discussion about what should be where: I totally agree with moving some code out of `core-executor` but I haven't yet understand the purpose of other new crates

rampage644 commented on crates/api-snowflake-rest/src/schemas.rs:
```diff
@@ -66,7 +66,7 @@ pub struct ClientEnvironment {
 
 #[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
-pub struct QueryRequest {
+pub struct AuthenticatedClient {
```
Can you explain the rename reasoning? `QueryRequest` -> `AuthenticatedClient` seems quite radical

YaroslavLitvinov replied:
I really have no clue about naming of a struct and a trait implementing interface for that struct.

rampage644 commented on crates/api-structs/Cargo.toml:
```diff
@@ -0,0 +1,27 @@
+[package]
+name = "api-structs"
```
Is it accurate to say `api-struct` is the crate that contains serialization/deserialization logic for UI component? If so, perhaps `api-ui-serde` makes more sense?

Or maybe this can be achieved with conditional compilation flag and better kept closer to `api-ui`?

YaroslavLitvinov replied:
Splitting traits by features. Serde is a feature, but then core-executor imports struct / type from i trait?

Second: Introducing flags you mentioned still keeps dependencies unchanged. What about separating of concerns?

rampage644 commented on crates/api-seed/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "api-seed"
```
Let;s have a discussion about this: if the only purpose of it is to seed data, perhaps it's better to have a python script seeding the data?

YaroslavLitvinov replied:
Honestly this PR could be divided into 2 parts: 
- HTTP Client implementation + Structs. 
- Seed, that is just a tool. But it can be a part of native executable or can external or both.

rampage644 commented on crates/api-client-rest/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "api-client-rest"
```
Is this just a Rust SDK to access embucket binary?

YaroslavLitvinov replied:
This is a client, that can be used as a library, or a separate executable. 
I don't know what is a Rust SDK, but this is a client which knows our communication protocol.

YaroslavLitvinov commented:
Code Review from Autopilot

Here's an overview of the code changes in PR #841 for Embucket/embucket (results may be incomplete due to result limits):

### Major Additions
- Introduced a new crate: `api-client-rest` (see new folder and Cargo.toml).
  - Implements an async client for REST API interactions, including authentication (login, refresh), and CRUD operations for volumes, databases, and schemas.
  - Includes modules for HTTP requests, error handling, and helper functions (e.g., cookie extraction).
  - Uses shared dependencies (`api-structs`, `async-trait`, `reqwest`, `serde`, etc.).

### Key File Changes
- **Cargo.toml / Cargo.lock**: Added new crates as workspace members and dependencies (notably `api-client-rest`, `api-seed`, `api-structs`). Updated dependencies like `reqwest`.
- **crates/api-client-rest/**:
  - `Cargo.toml`: Declares dependencies and crate info.
  - `src/api.rs`: Main API client logic and trait.
  - `src/lib.rs`: Exports modules.
  - `src/requests/client.rs`: Core HTTP logic, token management, and API request implementations.
  - `src/requests/error.rs`: Custom error types for HTTP and serialization.
  - `src/requests/helpers.rs`: Helpers for parsing cookies from response headers.
  - `src/requests/http.rs`: HTTP request wrappers over `reqwest`.
  - `src/requests/mod.rs`: Module organization.

### Notable Implementation Details
- The client uses Bearer tokens and cookies for authentication and session management.
- Implements proper error handling with custom enums.
- Uses async/await throughout for network operations.
- Modular structure: API logic, HTTP abstraction, error handling, and utilities are separated for maintainability.

### Suggestions / Questions
- The client interface and handling are well-structured, but integration tests would help ensure correctness.
- Consider documenting the expected usage patterns and any security considerations for token/cookie handling.
- Make sure all new workspace members are included in CI and tested.

View the full PR and all file diffs here: https://github.com/Embucket/embucket/pull/841/files

If you want to review a specific file or need a detailed breakdown of any section, let me know!
